jlm/mlir/frontend/MlirToRvsdg.cpp

```diff
@@ -0,0 +1,256 @@
+    return nullptr;
+  }
+  else if (mlirOperation.getName().getStringRef() == mlir::rvsdg::LambdaNode::getOperationName())
+  {
+    return convertLambda(mlirOperation, rvsdgRegion);
+  }
+  else if (mlirOperation.getName().getStringRef() == mlir::arith::ConstantIntOp::getOperationName())
+  {
+    auto constant = static_cast<mlir::arith::ConstantIntOp>(&mlirOperation);
+    auto type = constant.getType();
+    JLM_ASSERT(type.getTypeID() == mlir::IntegerType::getTypeID());
+    auto * integerType = static_cast<mlir::IntegerType *>(&type);
+
+    return rvsdg::node_output::node(
+        rvsdg::create_bitconstant(&rvsdgRegion, integerType->getWidth(), constant.value()));
+  }
+  else if (mlirOperation.getName().getStringRef() == mlir::rvsdg::LambdaResult::getOperationName())
+  {
+    // This is a terminating operation that doesn't have a corresponding RVSDG node
+    return nullptr;
+  }
+  else if (mlirOperation.getName().getStringRef() == mlir::rvsdg::OmegaResult::getOperationName())
+  {
+    // This is a terminating operation that doesn't have a corresponding RVSDG node
+    return nullptr;
+  }
+  else
+  {
+    throw util::error(
+        "Operation is not implemented:" + mlirOperation.getName().getStringRef().str() + "\n");
+  }
+}
+
+void
+MlirToRvsdg::convertOmega(mlir::Operation & mlirOmega, rvsdg::region & rvsdgRegion)
+{
+  // The Omega has a single region
+  JLM_ASSERT(mlirOmega.getRegions().size() == 1);
+  convertRegion(mlirOmega.getRegion(0), rvsdgRegion);
+}
+
+jlm::rvsdg::node *
+MlirToRvsdg::convertLambda(mlir::Operation & mlirLambda, rvsdg::region & rvsdgRegion)
+{
+  // Get the name of the function
+  auto functionNameAttribute = mlirLambda.getAttr(::llvm::StringRef("sym_name"));
+  JLM_ASSERT(functionNameAttribute != nullptr);
+  auto * functionName = static_cast<mlir::StringAttr *>(&functionNameAttribute);
+
+  // A lambda node has only the function signature as the result
+  JLM_ASSERT(mlirLambda.getNumResults() == 1);
+  auto result = mlirLambda.getResult(0).getType();
+
+  if (result.getTypeID() != mlir::rvsdg::LambdaRefType::getTypeID())
+  {
+    throw util::error("The result from lambda node is not a LambdaRefType\n");
+  }
+
+  // Create the RVSDG function signature
+  auto * lambdaRefType = static_cast<mlir::rvsdg::LambdaRefType *>(&result);
+  std::vector<std::unique_ptr<rvsdg::type>> argumentTypes;
+  for (auto argumentType : lambdaRefType->getParameterTypes())
+  {
+    argumentTypes.push_back(convertType(argumentType));
+  }
+  std::vector<std::unique_ptr<rvsdg::type>> resultTypes;
+  for (auto returnType : lambdaRefType->getReturnTypes())
+  {
+    resultTypes.push_back(convertType(returnType));
+  }
+  llvm::FunctionType functionType(std::move(argumentTypes), std::move(resultTypes));
+
+  auto rvsdgLambda = llvm::lambda::node::create(
+      &rvsdgRegion,
+      functionType,
+      functionName->getValue().str(),
+      llvm::linkage::external_linkage);
+
+  JLM_ASSERT(mlirLambda.getRegions().size() == 1);
```

Comment:
The `mlirLambda.getRegion()` does already an assert under the hood. No need for this `JLM_ASSERT` here.

Reply:
Removed